rfcs/mixed-atomicity-memory-model.md

```diff
@@ -0,0 +1,268 @@
+
+One question that came up in discussions with @OlivierNicole, @fabbing, @fpottier and @clef-men is whether we know how to describe the resulting model, where both atomic and non-atomic operations may be performed on a given location.
+
+The current RFC, prompted by a discussion with @clef-men, proposes a formal description of such a mixed-atomicity memory model.
+
+(Warning: we are not memory-model experts, so we are just making stuff up.
+cc @stedolan, @kayceesrk, @avsm, the authors of the initial model,
+and @gmevel, @jhjourdan and @fpottier as the authors of the Cosmo separation logic)
+
+
+## A detailed reminder on the current operational memory model
+
+In the current operational semantics for the memory-model, there are separate categories for atomic variables `A` and non-atomic variables `a`. The global store `S` stores a value `x` for each atomic location `A`, but for non-atomic locations `a` it stores a *history* `H`, a map from locations to values such that `H(t) = x` when a write `a := x` happened at timestamp `t`.
+
+A given thread has observed some writes to each non-atomic location `a`, but not necessarily all writes. When the thread reads from `a`, it reads non-determinstically the last-write value that it has observed, or any value more recent than that (written at a later timestamp). We call a *frontier* `F` a mapping from locations to timestamps that represents a time-of-last-observed-write for each location. Each thread has a frontier.
+
+Finally, the global store maps each atomic location `A` to a value `x`, but also to a fronter `Fᴬ`, that represents information propagated by threads that have written into this location. When a thread reads from a location `A`, it updates its own frontier `F` with the frontier of the location `Fᴬ` -- its frontier becomes the maximum of the timestamps, `F ⊔ Fᴬ`. When a thread *writes* to an atomic location, it updates its own frontier to become `F ⊔ Fᴬ`, but it also updates the atomic location's frontier in the same way.
+
+In LaTeX-turned-Unicode, this gives:
+
+```
+location              l
+atomic location       a
+non-atomic location   A
+value                 x
+timestamp             t ∈ Q (rational number)
+
+Machine     M := <S, P>
+Store       S := a ↦ H ⊎ A ↦ (F, x)
+History     H := t ↦ x
+Frontier    F := a ↦ t
+Program     P := i ↦ (F, e)
+transition  ϕ := read x | write x 
+
+e ~<l:ϕ>→ e'    S(l); F —<l:ϕ>→ C'; F'
+———————————————————————————————————————————————— Step
+<S, P || (F, e)]> —→ <S[l ↦ C'], P || (F', e')>
```

Comment:
Not sure to understand what the double pipe means here. Did you mean `P[i ↦ (F, e)]`?

Reply:
Yes, I strayed slightly from the paper here. In my mind `P || Q` is standard syntax for parallel composition of processes/threads, which avoids talking about the irrelevant thread index.

Reply:
We thought that the stray `]` may be a left-over from the previous notation, isn't it?